File: apps/api/src/application/commands/events/delete-event.command.ts
import { ICommand, ICommandHandler, CommandHandler } from '@nestjs/cqrs';
import {
  Inject,
  ForbiddenException,
  NotFoundException,
} from '@nestjs/common';
import {
  IEventRepository,
  EVENT_REPOSITORY,
} from '../../../domain/repositories/event.repository.interface';
import {
  ICoupleRepository,
  COUPLE_REPOSITORY,
} from '../../../domain/repositories/couple.repository.interface';

export class DeleteEventCommand implements ICommand {
  constructor(
    public readonly userId: string,
    public readonly coupleId: string,
    public readonly eventId: string,
  ) {}
}

@CommandHandler(DeleteEventCommand)
export class DeleteEventCommandHandler
  implements ICommandHandler<DeleteEventCommand>
{
  constructor(
    @Inject(EVENT_REPOSITORY)
    private readonly eventRepository: IEventRepository,
    @Inject(COUPLE_REPOSITORY)
    private readonly coupleRepository: ICoupleRepository,
  ) {}

  async execute(command: DeleteEventCommand): Promise<void> {
    const { userId, coupleId, eventId } = command;

    const couple = await this.coupleRepository.findById(coupleId);
    if (!couple || !couple.hasUser(userId)) {
      throw new ForbiddenException('User does not belong to this couple');
    }

    const event = await this.eventRepository.findById(eventId);
    if (!event) {
      throw new NotFoundException('Event not found');
    }

    if (!event.belongsToCouple(coupleId)) {
      throw new ForbiddenException('Event does not belong to this couple');
    }

    await this.eventRepository.delete(eventId);
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/application/command/event/DeleteEventCommand.kt
package com.couplecalendar.application.command.event

import com.couplecalendar.application.command.Command
import com.couplecalendar.application.command.CommandHandler
import com.couplecalendar.common.exception.ForbiddenException
import com.couplecalendar.common.exception.NotFoundException
import com.couplecalendar.domain.repository.CoupleRepository
import com.couplecalendar.domain.repository.EventRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

data class DeleteEventCommand(
    val userId: UUID,
    val coupleId: UUID,
    val eventId: UUID
) : Command<Unit>

@Component
class DeleteEventCommandHandler(
    private val eventRepository: EventRepository,
    private val coupleRepository: CoupleRepository
) : CommandHandler<DeleteEventCommand, Unit> {

    @Transactional
    override fun handle(command: DeleteEventCommand) {
        val couple = coupleRepository.findById(command.coupleId)
            ?: throw NotFoundException("Couple not found")

        if (!couple.hasUser(command.userId)) {
            throw ForbiddenException("User is not a member of this couple")
        }

        val event = eventRepository.findById(command.eventId)
            ?: throw NotFoundException("Event not found")

        if (!event.belongsToCouple(command.coupleId)) {
            throw ForbiddenException("Event does not belong to this couple")
        }

        eventRepository.delete(command.eventId)
    }
}
